Adds Elixir and Gleam as languages

kikofernandez opened the pull request:
#### What kind of change does this PR introduce?

Use the presence of property-based tests to detect the use of fuzzing in Elixir and Gleam code.

- [X] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

There is no support for fuzz detection for Elixir and Gleam.

#### What is the new behavior (if this is a feature change)?**

Adds Elixir and Erlang as a client language.
Scans for the import of property-based test frameworks to detect the use of fuzzing in Elixir and Gleam

- [X] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

NONE 
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.

Fixes #

or

NONE
-->

#### Special notes for your reviewer

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release,
include the string "ACTION REQUIRED".

For more information on release notes see: https://git.k8s.io/release/cmd/release-notes/README.md
-->

```release-note
Support detection of fuzzing in Elixir and Gleam through the import of property-based testing modules
```

kikofernandez commented on checks/raw/fuzzing_test.go:
```diff
@@ -394,6 +394,116 @@ func Test_checkFuzzFunc(t *testing.T) {
+			wantErr:  true,
+			langs: []clients.Language{
+				{
+					Name:     clients.Haskell,
```
Thanks a lot @voltone ! It was a good catch, because we were testing that no property-based testing happens, and based on the patterns, this surely should not have been detected by Haskell either :) 
Thanks!

spencerschrock commented on checks/raw/fuzzing.go:
```diff
@@ -85,6 +85,26 @@ var languageFuzzSpecs = map[clients.LanguageName]languageFuzzConfig{
+		Desc:        propertyBasedDescription("Elixir"),
+	},
+
+	// Fuzz patterns for Gleam based on property-based testing.
+	clients.Gleam: {
+		filePatterns: []string{"*.ex", "*.exs", "*.erl", "*.hrl"},
+		// Look for direct imports of PropCheck, and StreamData.
+		funcPattern: `(use\s+(PropCheck|ExUnitProperties)|` + // Elixir libraries
+			`-include_lib\("(eqc|proper)/include/(eqc|proper).hrl"\)\.|` + // Erlang libraries
+			`import\s+qcheck)`, // Gleam library
```
I'm unfamiliar with Gleam. Should it not be looking at a `.gleam` files?

kikofernandez replied:
Meaning, yes, I forgot to add `*.gleam` but Gleam also allows calling Erlang and Elixir files from within Gleam files.
I think @lpil and @giacomocavalieri can confirm this :)

spencerschrock replied:
And that wouldn't be covered by the existing Erlang/Elixir fuzz patterns? I worry these patterns will drift out of date if someone updates the patterns for Erlang or Elixir and forgets to update them here.

Scorecard will look at multiple languages in the repository (as defined by languages with more than 1/8th of the LOC in a repo). So I guess it depends on the ratio of files in a repo for this to matter or not?

kikofernandez replied:
Oh, you are right. If in a Gleam project scorecard detects Erlang or Elixir, then those will be recognise.
Absolutely right. I will fix this

kikofernandez replied:
I think this push should fix it. Thanks again, I somehow was thinking about all the combinations the language can do, instead of focusing on the files scorecard recognises. Thanks again!

kikofernandez replied:
> And that wouldn't be covered by the existing Erlang/Elixir fuzz patterns? I worry these patterns will drift out of date if someone updates the patterns for Erlang or Elixir and forgets to update them here.
> 
> Scorecard will look at multiple languages in the repository (as defined by languages with more than 1/(4*N)th of the LOC in a repo (where N = num langs). So I guess it depends on the ratio of files in a repo for this to matter or not?

Yes, it may depend. Maybe it makes sense to leave it for now specific to Gleam files, and if Gleam relies too much on Erlang or Elixir for property-based testing, maybe someone creates a new PR in that case?

It feels like, given that Gleam has its own property-based library, maybe the Gleam community uses that and does not rely that much on Erlang/Elixir for those tests.

Does this sound reasonable?

lpil replied:
A Gleam project maybe include a small amount of JavaScript or Erlang, but typically there is none.

It's similar to C in Python libraries.

kikofernandez replied:
Great, then this PR should be fine for most Gleam projects.
Thanks for the quick response rate @lpil  and @spencerschrock

spencerschrock commented on checks/raw/fuzzing_test.go:
```diff
@@ -394,6 +394,116 @@ func Test_checkFuzzFunc(t *testing.T) {
+			fileContent: "use ExUnitProperties",
+		},
+		{
+			name:     "Elixir with no property-based testing",
+			want:     false,
+			fileName: []string{"NoPropTest.exs"},
+			wantErr:  true,
```
we dont want `wantErr: true` here (I know this was copy/paste from Haskell, and you did the same in the Erlang PR. It wasn't caught in either of those reviews).

Also, the test was already written in a way that it will pass regardless of what wantErr is.

```diff
			name:     "Elixir with no property-based testing",
			want:     false,
			fileName: []string{"NoPropTest.exs"},
-			wantErr:  true,
```

spencerschrock commented on checks/raw/fuzzing_test.go:
```diff
@@ -394,6 +394,116 @@ func Test_checkFuzzFunc(t *testing.T) {
+			},
+			fileContent: "-include_lib(\"proper/include/proper.hrl\").",
+		},
+		{
+			name:     "Gleam with no property-based testing",
+			want:     false,
+			fileName: []string{"test.gleam"},
+			wantErr:  true,
```
```diff
		{
			name:     "Gleam with no property-based testing",
			want:     false,
			fileName: []string{"test.gleam"},
-			wantErr:  true,
```

kikofernandez commented:
Thanks for catching the issues in Elixir and Gleam. I have also corrected the bug for the Erlang commit and added it here, unless you prefer a specific PR for the Erlang broken test.

spencerschrock reviewed (approved):
Thanks for the contribution and quick responses! Please fix DCO and I can get this merged.

https://github.com/ossf/scorecard/pull/4408/checks?check_run_id=33054586665

kikofernandez commented on checks/raw/fuzzing.go:
```diff
@@ -85,6 +85,24 @@ var languageFuzzSpecs = map[clients.LanguageName]languageFuzzConfig{
+	clients.Gleam: {
+		filePatterns: []string{"*.gleam"},
+		// Look for direct imports of PropCheck, and StreamData.
+		funcPattern: `import\s+qcheck)`, // Gleam library
```
Thanks! This is what happens when I try to rush to work on this, but it's family dinner time.
Rushing stuff does not work for me :)

Thanks again

kikofernandez commented:
I have fixup (squash) some commits together and re-add the forgotten `signed-off-by`.
Also, I finally run tests locally, so this is the good final commit.

kikofernandez commented:
> Thanks for the contribution and quick responses! Please fix DCO and I can get this merged.
> 
> https://github.com/ossf/scorecard/pull/4408/checks?check_run_id=33054586665

@spencerschrock Thanks again for your fast response rate! 
I wished I had got it right the first time, and dinner time with kids did not help. But the end result is what matters :) 

Thanks again!

spencerschrock commented:
> I wished I had got it right the first time, and dinner time with kids did not help

No worries, family comes first :)
